### PR TITLE
Adds foldM

### PR DESCRIPTION
**Pull request by eed3si9n**
This implements `foldM`.

```scala
val names = List("Aaron", "Betty", "Calvin", "Deirdra")
val sumM = F.foldM(names, "") { (acc, x) => (Some(acc + x): Option[String]) }
assert(sumM == Some("AaronBettyCalvinDeirdra"))
val notCalvin = F.foldM(names, "") { (acc, x) => 
  if (x == "Calvin") (None: Option[String])
  else (Some(acc + x): Option[String]) }
assert(notCalvin == None)
```

### CODE REVIEW

File: tests/src/test/scala/cats/tests/FoldableTests.scala [hunk 1]
```diff
@@ -26,6 +26,12 @@ class FoldableTests extends CatsSuite {
     // more basic checks
     val names = List("Aaron", "Betty", "Calvin", "Deirdra")
     assert(F.foldMap(names)(_.length) == names.map(_.length).sum)
+    val sumM = F.foldM(names, "") { (acc, x) => (Some(acc + x): Option[String]) }
+    assert(sumM == Some("AaronBettyCalvinDeirdra"))
+    val notCalvin = F.foldM(names, "") { (acc, x) =>
```
Comment (ceedubs) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: Scalastyle doesn't like some trailing whitespace here.

**Comment by ceedubs** [comment]
This looks good to me once the whitespace is fixed and the build is passing.

@non could you please be a second set of eyes on the use of `partialFold`/`Fold`/`Lazy`, since you are the most familiar with that?

**Comment by eed3si9n** [comment]
Added the whitespace fix.

It's based on `foldRight` but I wasn't able to get it to not blow the stack, so I made it eager.

**Comment by non** [comment]
This looks good. I want to see if I can make it stack-safe, but in principle I'm :+1: on this.

**Comment by eed3si9n** [comment]
I suspect that to make it completely lazy, it's not enough to fold lazily, but also make `G.flatMap` part lazy, which involves trampolining it?

**Comment by ceedubs** [comment]
I'm a little hesitant on this one, now that I know it isn't stack-safe. If we think we could make it stack-safe in the future, I'm less concerned. Could we at least leave a scaladoc comment warning about stack safety?

**Comment by eed3si9n** [comment]
I think the issue was that I could make it lazy, but laziness only goes to calculating a `flatMap` of a `flatMap` of a `flatMap` etc, so when you put the value into it it blows the stack. So to fix that we need another datatype beyond `Lazy` like `Trampoline`?

**Comment by ceedubs** [comment]
@eed3si9n yes, I suspect that you are right.

**Comment by milessabin** [comment]
s/Trampoline/Eval/

**Comment by adelbertc** [comment]
This seems useful but it looks like it got lost in the PRs - very sorry @eed3si9n . If looks like there's some actual failing tests: https://travis-ci.org/typelevel/cats/builds/99910227 Once that's fixed this LGTM

**Comment by TomasMikula** [comment]
You seem to first build a function, that you then apply at once to the argument. Unless the target monad `G` is lazy (such as `Free`), the function you build is not stack-safe. To see this, let's view this piece of code

```
(k: B => G[B]) =>
        w: B => G.flatMap(f(w, a))(k)
```

as

  k<sub>i</sub> := w => G.flatMap(f(w, a<sub>i</sub>))(k<sub>i-1</sub>)

Thus, if `G.flatMap` is eager, then k<sub>n</sub> invokes k<sub>n-1</sub> invokes k<sub>n-2</sub> invokes ... invokes k<sub>0</sub>.

Now for the solution, I might be missing something, but wouldn't an eager `foldLeft` work?

**Comment by DavidGregory084** [comment]
@adelbertc I'm fairly certain this was submitted when FoldableTestsAdditional was flickering (see #769); it's possible that if you close and reopen the PR it will pass.

**Comment by eed3si9n** [comment]
Closing this in favor of #925

**Pull request closed without merging**